File: problem-solving/run-length-encoding/run-length-encoding.ts
const append = <T>(xs: T[]) => (element: T) => xs.push(element)

const getLast = (element: string) => element[element.length - 1]

export function runLengthEncoding(input: string): string {
  const chars: string[] = []
  let step = 1

  const appendToChars = append(chars)

  for (let i = 1; i < input.length; i++) {
    const current = input[i]
    const prev = input[i - 1]
    if (current !== prev || step === 9) {
      appendToChars(step.toString())
      appendToChars(prev)
      step = 0
    }
    step += 1
  }

  appendToChars(step.toString())
  appendToChars(getLast(input))

  return chars.join("")
}

const input = "AAAAAAAAAAAAABBCCCCDD"

console.log(runLengthEncoding(input))

File: problem-solving/run-length-encoding/run-length-encoding.go
package main

import (
	"fmt"
	"strconv"
)

func RunLengthEncoding(str string) string {
	chars := []byte{}
	step := 1

	for i := 1; i < len(str); i++ {
		current := str[i]
		prev := str[i-1]

		if current != prev || step == 9 {
			chars = append(chars, strconv.Itoa(step)[0])
			chars = append(chars, prev)
			fmt.Println(strconv.Itoa(step)[0])
			step = 0
		}
		step++
	}

	chars = append(chars, strconv.Itoa(step)[0])
	chars = append(chars, str[len(str)-1])

	return string(chars)
}

func main() {
	res := RunLengthEncoding("AAAAAAAAAAAAABBCCCCDD")
	fmt.Println(res)
}
